google/oauth2/service_account.py

```diff
@@ -429,6 +429,12 @@ def refresh(self, request):
                 "domain wide delegation is not supported for non-default universe domain"
             )
 
+        # If a refresh is triggered under the case of credential not expired
+        # that means some abnormal cases proactively trigger this refresh, we
+        # need to update the trust boundary cache.
+        if self.valid:
```

Comment:
I'm a bit confused, why does the trust boundary need to be refreshed here?

Reply:
The `refresh` method can be triggered by client code. In the library normal flow, we have 
https://github.com/googleapis/google-auth-library-python/blob/main/google/auth/credentials.py#L174-L175
```
        if not self.valid:
            self.refresh(request)
```
which is the flow of refresh in a case of credentials expired.

But one example of client `gcloud`, we are also do a `refresh` when the abnormal retryable case of authentication issue happens.

Reply:
but why do we want to reset the trust boundary in that case?

Reply:
That means an abnormal auth status happens, which could include trust boundary error. We only do the reset trust boundary in that case. Otherwise, a trust boundary value will be carried from the creating of credential object till the entire lifecycle end.